### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by carlosEGuerra**

### CODE REVIEW

File: Assignment_Week_3.cs [hunk 1]
```diff
@@ -22,6 +22,7 @@ public static int[][] uniqueTripletCombinations(int[] userInput)
             //organizes the user input
             Array.Sort(userInput);
 
+            //O(n^3) Complexity, this REALLY does need to be improved
```
Comment (itsnotdevin) [inline Assignment_Week_3.cs]: # Hi @carlosEGuerra 
Thank you for sharing your solution! I really enjoyed looking through it! I have some insight on how to optimize this homework assignment, so I wrote an explanation using similar variable names for your consideration.  The suggested changes should bring the run-time complexity to ~O(n^2), but increases the space complexity by ~O(n). I wrote the code in Java, but the syntax is very similar and the ideas are the same.  Please take a peek below, and I hope you find this information useful!  :)

- **_Motivation:_** We can modify the solution approach to achieve O(n^2) run-time complexity. These improvements potentially outweigh the increase in space complexity.
- _**Idea:**_ We can use _(-1 * input[]), a HashSet, and the 2Sum solving technique to remove sorting and the third nested loop. 
- _**Goals:**_ O(n^2) run-time complexity. (Woohoo!)
- _**Considerations:**_ This approach adds O(n) space complexity to the current solution approach.  

## An Example
- input[first] = {**-2**, 1, 1};
- target = **2**;
```
     Integer target = 0;
     HashSet<Integer> hash = new HashSet<Integer>();
     ArrayList<Integer[]> list = new ArrayList<Integer[]>();
     
     // The value at the first index, input[first] is -2.
     // so our target value is now (-1 * - 2) = 2
     for (int first = 0; first < input.length; first++)  {
          target = (-1 * input[first]);
```
We begin by initializing our local variables, and start iterating over the length of the input array. For each iteration, calculate a value that equals zero when added to input[index].  Since we're looking for valid tuple combinations that include the value at input[index], the other two numbers must add up to this target value for the three to add up to zero.
_**NOTE:** During implementation, you could reference the array position directly by using input[first] instead of creating a target Integer._

- input[first] = {**-2**, 1, 1};
- target = **2**;
- input[second] = {-2, **1**, 1};
```
     // second and first indices are equal, so we continue
     // second++, input[second] = 1
     for (int second = 0; second < input.length; second++) {
          if (second == first) {
               continue;
          }
```
We then start iterating over the input array with the second index.  If the first and second index are the same, we need to continue to avoid considering the same Integer twice. 
_**NOTE:** This is the same technique that can be used to solve 2Sum._

- input[first] = {**-2**, 1, 1};
- target = **2**;
- input[second] = {-2, **1**, 1};
- hash = { };
- list = { };
```
          // Our hashset does not contain input[second] = 1  
          // So, we add the difference between the target and input[second] to the hashset. 
          // 2-1 = 1 hash.add(1);    
          if (hash.contains(input[second]) {
               ...
               }
          }
         hash.add(target-input[second]);
```
_If input[second] is NOT in our hashset:_ We add the difference between the target and input[second] to the hashset.  When we encounter this value later during our iteration, we know we have found a value pair who's sum equals the target value (making the whole tuple add up to zero).

- input[first] = {**-2**, 1, 1};
- target = **2**;
- input[second] = {-2, 1, **1**};
- hash = {**1**};
- list = { };
```
          // input[second] = 1;  We find the value in our hash.
          // tmp[0] = -2, tmp[1] = 1, tmp[2] = 2-1=1 
          // list does not contain this array, so we add it to the list.
          if (hash.contains(input[second]) {
               Integer[] tmp = new Integer[3];
               tmp[0] = input[first];
               tmp[1] = input[second];
               tmp[2] = (target - input[second]);
               if (!list.contains(tmp)) {
                    list.Add(tmp);
               }
               ...
```
_If input[second] is in our hash:_ Bingo! We found the value pair that adds up to the target value. We know input[first] = -2 is a part of the solution array. We also know the input[second] is a part of our solution array.  We can figure out the third tuple value by calculating the difference between the target value and input[second].  We verify we haven't already added this tuple, and then add it to our list.

- input[first] = {**-2**, 1, 1};
- target = **2**;
- input[second] = {-2, 1, **1**};
- hash = { };
- list = {[-2,1,1]};
```
          hash.clear();
```   
Once the second index has completed iterating, we need to empty our hash in preparation for the next tuple evaluation with input[first]++.  Once both indices have completed iterating, we prepare and return our list.

## The Code
- list = {[-2,1,1]};
- input[] = {-2, 1, 1};

```
     Integer target = 0;
     HashSet<Integer> hash = new HashSet<Integer>();
     ArrayList<Integer[]> list = new ArrayList<Integer[]>();

     for (int first = 0; first < input.length; first++)  
     {
          target = (-1 * input[first]);
          for (int second = 0; second < input.length; second++) 
          {
               if (second == first)
               {  
                    continue;
               }
               if (hash.contains(input[second]) 
               {
                    Integer[] tmp = new Integer[3];
                    tmp[0] = input[first];
                    tmp[1] = input[second];
                    tmp[2] = (target - input[second]);
                    if (!list.contains(tmp)) 
                    {
                         list.Add(tmp);
                    }
               }
               hash.add(target-input[second]);
          }
          hash.clear();
     }
     return list.ToArray();
```

**Review by itsnotdevin — COMMENTED** [review]
Optimization Suggestion

File: Assignment_Week_3.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            //if there isn't enough input
+            if (userInput.Length < 3)
+                throw new ArgumentException();
+
+            //used to contain the pending results
+            List<int[]> results = new List<int[]>();
+
+            //organizes the user input
+            Array.Sort(userInput);
+
```
Comment (loresfca) [inline Assignment_Week_3.cs:24]: Hey! maybe here you don't need to sort the array because it doesn't impact your solution.

File: Assignment_Week_3.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            //O(n^3) Complexity, this REALLY does need to be improved
```
Comment (loresfca) [inline Assignment_Week_3.cs:25]: An idea to improve your solution may be to implement some sort of memoization :blush:, for example, you could implement a set to know if you have already checked a number. Also, twoSum could be good place to start.

File: Assignment_Week_3.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                            temp[1] = userInput[secondIndex];
+                            temp[2] = userInput[thirdIndex];
+                            if(!results.Contains(temp))
+                                results.Add(new int[] { userInput[firstIndex], userInput[secondIndex], userInput[thirdIndex]});
+                        }
+                    }
+                }
+            }
+
+            return results.ToArray();
```
Comment (loresfca) [inline Assignment_Week_3.cs:45]: Here you could return a list directly :relaxed:

File: Assignment_Week_3.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+               for(int secondIndex = firstIndex + 1; secondIndex < userInput.Length - 2; secondIndex++)
+                {
+                    for(int thirdIndex = secondIndex + 1; thirdIndex < userInput.Length - 1; thirdIndex++)
+                    {
+                        if((userInput[firstIndex] + userInput[secondIndex] + userInput[thirdIndex]) == 0)
+                        {
+                            int[] temp = new int[3];
+                            temp[0] = userInput[firstIndex];
+                            temp[1] = userInput[secondIndex];
+                            temp[2] = userInput[thirdIndex];
```
Comment (loresfca) [inline Assignment_Week_3.cs:37]: What happens if try to add a triplet with the same values but the order is different?

File: Assignment_Week_3.cs (hunk 4 above)
Comment (charientism) [reply Assignment_Week_3.cs:45]: Either is good :) Always remember to ask your interviewer what the return type should be.

File: Assignment_Week_3.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            Array.Sort(userInput);
+
+            //O(n^3) Complexity, this REALLY does need to be improved
+            for(int firstIndex = 0; firstIndex < userInput.Length - 3; firstIndex++)
+            {
+               for(int secondIndex = firstIndex + 1; secondIndex < userInput.Length - 2; secondIndex++)
+                {
+                    for(int thirdIndex = secondIndex + 1; thirdIndex < userInput.Length - 1; thirdIndex++)
+                    {
+                        if((userInput[firstIndex] + userInput[secondIndex] + userInput[thirdIndex]) == 0)
```
Comment (charientism) [inline Assignment_Week_3.cs:32]: Solid brute-force implementation. Good job recognizing that 1) the Big O can be improved further via optimization and 2) accounting for duplicates by sorting your array beforehand. We'll go over how to do reduce the Big O tomorrow :)

File: UnitTest1.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+{
+    [TestClass]
+    public class UnitTest1
+    {
+        private int[] inputArrayForTest = new int[6];
+        private int[][] testArray1 = new int[2][];
+
+        [TestInitialize]
+        public void arrayInitializationForTesting()
+        {
```
Comment (charientism) [inline UnitTest1.cs:16]: Populating the array this way is definitely valid. If you're looking to save space while maintaining readability, you can also do the following: inputArrayForTest = {-1, 0, 1, 2, -1, -4};

File: UnitTest1.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            int[] temp2 = new int[3];
+            temp2[0] = -1;
+            temp2[1] = -1;
+            temp2[2] = 2;
+
+            testArray1[0] = temp1;
+            testArray1[1] = temp2;
+
+        }
+        [TestMethod]
```
Comment (charientism) [inline UnitTest1.cs:40]: This is a good test case for validating that your code works on duplicate numbers and can find multiple solutions. Some other ones that would be good to consider are edge cases (null, arrays with less than three elements, etc.).

File: Assignment_Week_3.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
```
Comment (charientism) [inline Assignment_Week_3.cs:4]: Thanks for uploading your solution/test case! Going forward, we're also hoping to give you guys feedback on your pen/paper TEBOW IT, so make sure to upload that as well :)

**Review by charientism — APPROVED** [review]
